### ISSUE TITLE
Field is not a Ring

### ISSUE THREAD

**Issue by andriusstank** [tonyday567/numhask#139]
I was surprised to find out subtraction doesn't work on `Field`s.

I see `Distributed` replaced `CommutativeRing` as a superclass of `IntegralDomain` in a707fb287c8608de43385459c49fded6b5b19c78. I can't tell if it was intentional or not, but I think if it wasn't meant to be a subclass of ring, it shouldn't be named "Field".

In the meantime, svg diagram is outdated - `CRing` was renamed to `CommutativeRing`. Also the arrow CRing -> IntegralDomain is incosistent with code, which is the subject of this issue.

**Comment by tonyday567** [comment]
Thanks for the report!

I'm surprised that subtraction dropped out as well! I turned my mind off following ghc hints and Subtractive disappeared as the hedgehog tests make this implicit due to the random generation phase.

Would you agree that Field is (+,-,*,/)? Specifically, that it doesn't need to be related to rings in any way?

**Comment by tonyday567** [comment]
#140 and 4ae52837cdc79376d95b0e44e587868d05050208

**Comment by andriusstank** [comment]
I don't feel knowledgeable enough to suggest how to model abstract algebra concepts in Haskell. It's a very nuanced problem.

A field is a ring. So I think it's important that `Field` constraint implies `Ring` constraint. That is, if I have a function over ring

    f :: Ring a => ...

then I can use it for fields:

    g :: Field a => ...
    g ... = ... f ...

This works. I guess it's good as it is, there is no need to have `Ring` as a superclass of `Field`.

**Comment by tonyday567** [comment]
If you try and make the class structure vertically hierarchical, you soon run into this warning from GHC:

```
[-Wsimplifiable-class-constraints]
    • The constraint ‘CommutativeRing a’
        matches an instance declaration
      instance (Distributive a, Subtractive a) => CommutativeRing a
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the context: (CommutativeRing a, Distributive a,
                       Subtractive a, Divisive a)
      While checking the super-classes of class ‘Field’
      In the class declaration for ‘Field’
```
I'm not exactly sure about the details, but it has something to do with whether let clauses float up in core, and I make it a habit to take GHC warnings seriously when it comes to core efficiencies.

Other than this I agree with you about CommutativeRing, and the new diagram reflects this, even if it isn't reflected in the class structure.

**Comment by andriusstank** [comment]
Looking at `CommutativeRing` now I see it is equivalent to `Ring` which feels wrong. It's almost a synonym, as every ring is granted an instance automatically.

Speaking of `Ring`, can't you just enable `MonoLocalBinds` to silence the warning?

Another option is 

    type Ring a = (Distributive a, Subtractive a)

This required `ConstraintKinds` and you won't see `Ring` in inferred types, but I can't see other downsides.

**Comment by LeanderK** [comment]
While

```
type Ring a = (Distributive a, Subtractive a)
```

is possible, it's rarely used in practice since unfortunately type-synonym don't get the same treatment as type-classes. The class-instances trick is used quite often (there was a blog-post detailing it, but right now I only find this reddit-answer: https://www.reddit.com/r/haskell/comments/ab8ypl/monthly_hask_anything_january_2019/edk1ot3/)

Btw. not every ring is commutative and they occur quite often in practice (for example the square matrices!).

**Comment by tonyday567** [comment]
CommutativeRing was separate from Ring because a new law was introduced.  There isn't a good solution to actually adding Commutative, however, so I agree with you that they are effectively the same, and I dropped CommutativeRing in the latest push.  New diagram looks pretty neat now.

So far, numhask is a zero-cost abstraction: the speed is the same before and after all this class guff. I have turned on MonoLocalBinds in the past but i worry that it might cost us somewhere one day.
